R/cube-result.R

```diff
@@ -388,6 +411,122 @@ setMethod("margin.table", "CrunchCube", function (x, margin=NULL) {
+    
+    # if there are any _items in the margin to collapse, be wary!
+    collapsed_items <- endsWith(getDimTypes(x)[margins_to_collapse], "_items")
+
+    # grab the old cube structure and over-write it with new subset data.
+    out <- x
+    
+    # iterate through measures, collapsing the dimension(s) specified.
+    out@arrays[] <- lapply(out@arrays, function(arr){
+        # off_margins here are the margins that we want to keep, however because
+        # we treat items dimensions differently, we need to add them to the
+        # off_margins, and deal with them after we collapse by summing
+        off_margins <- margins_to_keep
+        if (any(collapsed_items)) {
+            # mean across any items dimension
+            items_to_collapse <- margins_to_collapse[collapsed_items]
+            off_margins <- sort(c(off_margins, items_to_collapse))
+        }
+        
+        # collapse the margins to collapse by summing across them
+        array <- as.array(apply(arr, off_margins, sum))
+
+        # now, we need to get rid of the extra items dimensions that we want to
+        # collapse across. We cannot simply sum across them because they
+        # represent more than one observation per row; and we don't want to
+        # double/triple/etc. count those observations. Instead we take the mean.
+        # Another option would be to just take the unique value or first value
+        # (since they should all be the same at this point given that they have
+        # been collapsed above), however there are tiny differences due to
+        # floating-point precision and weighting in the calculation in ZZ9.
+        if (any(collapsed_items)) {
+            # mean across any items dimension
+            result_dim <- seq_along(dim(array))
+            final_names <- dimnames(array)[which(!off_margins %in% items_to_collapse)]
+            array <- as.array(apply(array, setdiff(result_dim, which(off_margins %in% items_to_collapse)), mean, na.rm = TRUE))
+            dimnames(array) <- final_names
+        }
+        
+        # add attributes
+        attributes(array)$variable$type <- attributes(arr)$variable$type
+        
+        return(array)
+    })
+    # we needed to include the selected dim of MRs in off_margins above, but
+    # when subsetting dimensions we definitely do not want them to be included.
+    out@dims <- out@dims[margins_to_keep]
+    out$query$dimensions <- out$query$dimensions[margins_to_keep]
+    out$result$dimensions <- out$result$dimensions[margins_to_keep]
+
+    # Need to calculate the total missing to fill in, to do this, set all
+    # non-missing cells to zero, and then sum across all dimensions (with the
+    # exception of `_items` dimensions)
+    unweighted_counts <- as.array(out@arrays$.unweighted_counts)
+    non_missing <- evalUseNA(unweighted_counts, out@dims, "no")
+    all_missings <- do.call("[<-", c(list(x=unweighted_counts), non_missing, value = 0))
```

Comment:
Something about this makes me a bit uncomfortable, maybe just because I don't associate assignment with summing, is there a way to accomplish this which doesn't rely on a `do.call`?